File: chrome/browser/finds/android/finds_feature_map.cc
// Copyright 2026 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "base/android/feature_map.h"
#include "base/no_destructor.h"
#include "chrome/browser/finds/android/jni_headers/FindsFeatureMap_jni.h"
#include "chrome/browser/finds/core/finds_features.h"

namespace finds::android {

static int64_t JNI_FindsFeatureMap_GetNativeMap(JNIEnv* env) {
  static const base::Feature* const kFeaturesExposedToJava[] = {
      &features::kChromeFinds};
  static base::NoDestructor<base::android::FeatureMap> kFeatureMap(
      kFeaturesExposedToJava);
  return reinterpret_cast<int64_t>(kFeatureMap.get());
}

}  // namespace finds::android

DEFINE_JNI(FindsFeatureMap)

File: chrome/browser/finds/android/java/src/org/chromium/chrome/browser/finds/FindsFeatureMap.java
// Copyright 2026 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.chrome.browser.finds;

import org.jni_zero.JNINamespace;
import org.jni_zero.NativeMethods;

import org.chromium.base.FeatureMap;
import org.chromium.build.annotations.NullMarked;
import org.chromium.build.annotations.Nullable;

/** Maps native finds features to java. */
@JNINamespace("finds::android")
@NullMarked
public class FindsFeatureMap extends FeatureMap {
    private static @Nullable FindsFeatureMap sInstance;

    // Not directly instantiable.
    private FindsFeatureMap() {
        super();
    }

    /**
     * @return the singleton FindsFeatureMap.
     */
    public static FindsFeatureMap getInstance() {
        if (sInstance == null) sInstance = new FindsFeatureMap();
        return sInstance;
    }

    @Override
    protected long getNativeMap() {
        return FindsFeatureMapJni.get().getNativeMap();
    }

    @NativeMethods
    protected interface Natives {
        long getNativeMap();
    }
}
